dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -57,42 +62,106 @@ public WorkerLocationPolicy(int numVirtualNodes) {
   public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWorkerInfos,
                                                    String fileId,
                                                    int count) {
-    if (blockWorkerInfos.size() == 0) {
+    if (blockWorkerInfos.isEmpty()) {
       return ImmutableList.of();
     }
     HASH_PROVIDER.refresh(blockWorkerInfos, mNumVirtualNodes);
     return HASH_PROVIDER.getMultiple(fileId, count);
   }
 
-  private static class ConsistentHashProvider {
+  static class ConsistentHashProvider {
     private static final HashFunction HASH_FUNCTION = murmur3_32_fixed();
-    private static final int MAX_ATTEMPTS = 100;
-    private List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
-    private NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
-
-    private volatile long mLastUpdatedTimestamp = 0L;
-
-    private final AtomicBoolean mNeedUpdate = new AtomicBoolean(false);
-
-    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 1000L;
+    private final int mMaxAttempts;
+    private final long mWorkerInfoUpdateIntervalNs;
+
+    /**
+     * Timestamp of the last update to {@link #mActiveNodesByConsistentHashing}.
+     * Must use System.nanoTime to ensure monotonic increment. Otherwise, updates to the worker
+     * list may be missed as the expiry based on TTL cannot be reliably determined.
+     */
+    private final AtomicLong mLastUpdatedTimestamp = new AtomicLong(System.nanoTime());
+    /**
+     * Counter for how many times the map has been updated.
+     */
+    private final AtomicLong mUpdateCount = new AtomicLong(0);
+    /**
+     * The worker list which the {@link #mActiveNodesByConsistentHashing} was built from.
+     * Must kept in sync with {@link #mActiveNodesByConsistentHashing}.
+     * Used to compare with incoming worker list to skip the heavy build process if the worker
+     * list has not changed.
+     */
+    private volatile List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
+    /**
+     * A map of virtual node indices to the actual workers.
+     * This is lazily initialized, can only be null before the first call to refresh.
+     */
+    @Nullable
+    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
```

Comment:
annotation?
```suggestion
    @GuardedBy("mActiveNodesByConsistentHashing")
    private volatile NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
```

Reply:
It's not clear which lock object to put in the `GuardedBy` annotation. The list is protected by a synchronized lock during initialization. After initialization, it's a volatile variable, and access is synchronized by CAS on an atomic timestamp.

Reply:
My best try would be a comment like above. Maybe we don't need `@GuardedBy` as long as the javadoc here is clear enough.